[Clang][Comments] Support for parsing headers in Doxygen \par commands

hdoc opened the pull request:
### Background

Doxygen's `\par` command ([link](https://www.doxygen.nl/manual/commands.html#cmdpar)) has an optional argument, which denotes the header of the paragraph started by a given `\par` command.

In short, the paragraph command can be used with a heading, or without one. The code block below shows both forms and how the current version of LLVM/Clang parses this code:
```
$ cat test.cpp
/// \par User defined paragraph:
/// Contents of the paragraph.
///
/// \par
/// New paragraph under the same heading.
///
/// \par
/// A second paragraph.
class A {};

$ clang++ -cc1 -ast-dump -fcolor-diagnostics -std=c++20 test.cpp
`-CXXRecordDecl 0x1530f3a78 <test.cpp:11:1, col:10> col:7 class A definition
  |-FullComment 0x1530fea38 <line:2:4, line:9:23>
  | |-ParagraphComment 0x1530fe7e0 <line:2:4>
  | | `-TextComment 0x1530fe7b8 <col:4> Text=" "
  | |-BlockCommandComment 0x1530fe800 <col:5, line:3:30> Name="par"
  | | `-ParagraphComment 0x1530fe878 <line:2:9, line:3:30>
  | |   |-TextComment 0x1530fe828 <line:2:9, col:32> Text=" User defined paragraph:"
  | |   `-TextComment 0x1530fe848 <line:3:4, col:30> Text=" Contents of the paragraph."
  | |-ParagraphComment 0x1530fe8c0 <line:5:4>
  | | `-TextComment 0x1530fe898 <col:4> Text=" "
  | |-BlockCommandComment 0x1530fe8e0 <col:5, line:6:41> Name="par"
  | | `-ParagraphComment 0x1530fe930 <col:4, col:41>
  | |   `-TextComment 0x1530fe908 <col:4, col:41> Text=" New paragraph under the same heading."
  | |-ParagraphComment 0x1530fe978 <line:8:4>
  | | `-TextComment 0x1530fe950 <col:4> Text=" "
  | `-BlockCommandComment 0x1530fe998 <col:5, line:9:23> Name="par"
  |   `-ParagraphComment 0x1530fe9e8 <col:4, col:23>
  |     `-TextComment 0x1530fe9c0 <col:4, col:23> Text=" A second paragraph."
  `-CXXRecordDecl 0x1530f3bb0 <line:11:1, col:7> col:7 implicit class A
```

As we can see above, the optional paragraph heading (`"User defined paragraph"`) is not an argument of the `\par` `BlockCommandComment`, but instead a child `TextComment`.

For documentation generators like [hdoc](https://hdoc.io/), it would be ideal if we could parse Doxygen documentation comments with these semantics in mind. Currently that's not possible.

### Change

This change parses `\par` command according to how Doxygen parses them, making an optional header available as a an argument if it is present. In addition:

- AST unit tests are defined to test this functionality when an argument is present, isn't present, with additional spacing, etc.
- TableGen is updated with an `IsParCommand` to support this functionality
- `lit` tests are updated where needed

hdoc committed:
Support for parsing headers in Doxygen \par commands

AaronBallman commented on clang/lib/AST/CommentParser.cpp:
```diff
@@ -149,6 +149,63 @@ class TextTokenRetokenizer {
 
+  /// Check if this line starts with @par or \par
+  bool startsWithParCommand() {
+    unsigned Offset = 1;
+
+    /// Skip all whitespace characters at the beginning.
+    /// This needs to backtrack because Pos has already advanced past the
+    /// actual \par or @par command by the time this function is called.
+    while (isWhitespace(*(Pos.BufferPtr - Offset)))
+      Offset++;
```
You can use `StringRef::ltrim()` to handle this instead of doing it manually.

hdoc replied:
In this case we actually have to read the buffer backwards to find the `@par`, past `Pos.BufferStart` so it's a bit inconvenient to construct the `StringRef` which starts at some unknown position before `Pos.BufferStart`. In theory someone could have put 400 spaces between `@par` and the comment text, and we'd have to accommodate that when constructing the `StringRef`. 

I think the solution as it's written right now handles that well enough. I could make it count spaces or look up the previous token containing the `@par` and compute the correct StringRef for it, but I think this is good enough.

AaronBallman replied:
Okay, I think what's got me confused is that you calculate `Offset` but never actually use it (or modify `Pos.BufferPtr`). I was thinking you would have been able to use `ltrim().starts_with(...)` but I see now that the issue is we're at the start of the comment, beyond seeing `par`, and we want to backtrack.

But the logic below seems to be that we already know we're parsing a par command, so why do we need to do this check in the first place? (e.g., we get into this code path via:
```
 else if (Info->IsParCommand)
      S.actOnBlockCommandArgs(BC,
                              parseParCommandArgs(Retokenizer, Info->NumArgs));
```
and eventually land here, so don't we already know this is a par command?)

hdoc replied:
The check is needed because Doxygen interprets the `@par` command differently if there is text on the same line or not.

If there is text on the same line, such as in Example 1 below, then the text is interpreted to be the paragraph heading, and that heading is stored as an argument for the comment. If there is no text on the same line, i.e. `@par` is alone on the line as in Example 2, then the paragraph is interpreted to have no heading.

```
/// @par I am a paragraph heading
void example1() {}

/// @par 
/// I am a paragraph with no heading
void example2() {}
```

To be able to correctly parse the paragraph heading, the lexer needs to go and check where the `@par` actually is, because if it's on the same line as some text we need to slurp that text and put it into the Arg for the comment. This code implements the functionality for this check by backtracking until it sees whitespace, and then checking whether the text immediately preceding that whitespace is a `@par` command or not.

All that being said, your feedback is helpful and I did find a small bug while re-checking this code. `Offset` should be used, otherwise the code has an implicit assumption that there is only one `char` of whitespace between the text and the `@par`. I've fixed that in the latest commit. Thank you for pointing that out :)

AaronBallman replied:
Rather than backtracking, can't we still do a peek ahead? e.g., once we've seen `@par`, we can do a peek to see if the next token is a newline or not (or do a scan through the buffer to see if there is a non-whitespace character or only whitespace ending with a newline/eof)?

hdoc replied:
The issue is that the position of `Pos.BufferPtr` is after the `@par` command when `lexParHeading` is called. I believe this is because the comment parser [consumes](https://github.com/llvm/llvm-project/blob/84dd803993fd2b6b31f8168a3f4dc729406bd3ca/clang/lib/AST/CommentParser.cpp#L407) the `@par` command token itself during parsing, and then based on the token [it decides to call](https://github.com/llvm/llvm-project/blob/84dd803993fd2b6b31f8168a3f4dc729406bd3ca/clang/lib/AST/CommentParser.cpp#L434-L436) `lexParHeading`. I've done some tests locally to verify this behavior, and my observations align with what I describe above.

While I agree that it'd be preferable to do a readahead instead of backtracking, I don't see any easy ways to refactor the PR to incorporate this behavior as it currently stands. It looks like we'd have to refactor a little more of the comment parser, which is delicate and would incorporate changes outside of the scope of this change.

AaronBallman commented on clang/lib/AST/CommentParser.cpp:
```diff
@@ -149,6 +149,63 @@ class TextTokenRetokenizer {
+      if (Pos.BufferPtr + 1 == Pos.BufferEnd) {
+        consumeChar();
+        break;
+      } else {
+        consumeChar();
+      }
```
```suggestion
      }
      consumeChar();
```
No need for an `else` because we'd have broken out of the loop if we entered the `if`.

hdoc replied:
Fixed

AaronBallman commented on clang/lib/AST/CommentParser.cpp:
```diff
@@ -149,6 +149,63 @@ class TextTokenRetokenizer {
+      }
+    }
+
+    const unsigned Length = WordText.size();
```
```suggestion
    unsigned Length = WordText.size();
```
(We don't really use top-level `const`.)

hdoc replied:
There are a few other instances of it in this file, so originally I elected to follow the file-local convention. However, I have accepted your change

AaronBallman commented on clang/lib/AST/CommentParser.cpp:
```diff
@@ -304,6 +361,23 @@ Parser::parseCommandArgs(TextTokenRetokenizer &Retokenizer, unsigned NumArgs) {
 
+ArrayRef<Comment::Argument>
+Parser::parseParCommandArgs(TextTokenRetokenizer &Retokenizer,
+                            unsigned NumArgs) {
```
Can we assert that `NumArgs > 0`?

hdoc replied:
Fixed

hdoc committed:
Address review feedback

hdoc committed:
Small fix to par heading check and comments

hdoc commented:
Ping. This is rebased on latest main (which integrates the other comment parser changes we've made) and should be ready to go.

hdoc committed:
Run clang-format over failing file

Not sure why a failure is being triggered here as we didn't touch this
exact section. Our changes affect code a little further down in the
file.

hdoc committed:
More clang-format fixes

hdoc committed:
Clang format fix

AaronBallman reviewed (approved):
LGTM! Please be sure to add a release note to clang/docs/ReleaseNotes.rst so users know about the new functionality.

hdoc committed:
Add release notes blurb for \par command comment arg support

hdoc commented:
@AaronBallman thanks for the reminder :) the release notes blurb has been added
